compiler/rustc_borrowck/src/region_infer/opaque_types.rs

```diff
@@ -396,17 +398,14 @@ fn check_opaque_type_parameter_valid(
         } else {
             // Prevent `fn foo() -> Foo<u32>` from being defining.
             let opaque_param = opaque_generics.param_at(i, tcx);
-            tcx.sess
-                .struct_span_err(span, "non-defining opaque type use in defining scope")
-                .span_note(
-                    tcx.def_span(opaque_param.def_id),
-                    &format!(
-                        "used non-generic {} `{}` for generic parameter",
-                        opaque_param.kind.descr(),
-                        arg,
-                    ),
-                )
-                .emit();
+            tcx.sess.emit_err(OpaqueTypeNotDefine {
+                cause: OpaqueTyDefineErrCause::NonGenericUsed {
+                    span: tcx.def_span(opaque_param.def_id),
+                    descr: opaque_param.kind.descr().to_string(),
```

Comment:
Using `descr()` isn't ideal because it includes an English string through a interpolated variable.